File: asgate-js/asgate-js/test/oauth.test.ts
import { describe, expect, it } from 'vitest'
import { createClient } from '../src/asgate-client'
import { fakeJwt, createMockFetch } from './helpers'

function envelope(data: unknown) {
  return { status_code: 200, message: 'ok', data }
}

describe('AsgateOAuthApi', () => {
  it('getOidcSignInUrl devuelve provider y authorize_url con slug header', async () => {
    const { fetchImpl, calls } = createMockFetch((url) => {
      if (url.endsWith('/api/v1/auth/oidc/start')) {
        return {
          status: 200,
          body: envelope({
            provider: 'oidc_custom',
            authorize_url: 'https://idp.example.com/authorize?client_id=x',
          }),
        }
      }
      return { status: 404, body: {} }
    })
    const client = createClient({
      url: 'http://localhost:4440',
      organizationSlug: 'acme',
      fetch: fetchImpl,
    })
    const res = await client.oauth.getOidcSignInUrl()
    expect(res.provider).toBe('oidc_custom')
    expect(res.authorizeUrl).toContain('https://idp.example.com')
    const call = calls.find((c) => c.url.endsWith('/api/v1/auth/oidc/start'))
    const headers = call?.init.headers as Record<string, string>
    expect(headers['X-Organization-Slug']).toBe('acme')
  })

  it('handleOidcCallback intercambia code/state y guarda sesión', async () => {
    const accessToken = fakeJwt({ sub: 'u1', email: 'a@b.c', exp: 2000000000 })
    const { fetchImpl } = createMockFetch((url) => {
      if (url.includes('/api/v1/auth/oidc/callback')) {
        return {
          status: 200,
          body: envelope({
            redirect_url: `https://app/auth/callback#access_token=${accessToken}&refresh_token=ref`,
          }),
        }
      }
      return { status: 404, body: {} }
    })
    const client = createClient({
      url: 'http://localhost:4440',
      organizationSlug: 'acme',
      fetch: fetchImpl,
    })
    const res = await client.oauth.handleOidcCallback(
      'https://app/auth/callback?code=abc&state=xyz',
    )
    expect(res.session?.refreshToken).toBe('ref')
    expect(res.session?.user?.id).toBe('u1')
    expect(client.isSignedIn).toBe(true)
  })

  it('handleOidcCallback lanza si falta code o state', async () => {
    const { fetchImpl } = createMockFetch(() => ({ status: 200, body: {} }))
    const client = createClient({
      url: 'http://localhost:4440',
      organizationSlug: 'acme',
      fetch: fetchImpl,
    })
    await expect(
      client.oauth.handleOidcCallback('https://app/auth/callback?code=abc'),
    ).rejects.toThrow('missing code or state')
  })

  it('handleOidcCallback lanza si el query trae error', async () => {
    const { fetchImpl } = createMockFetch(() => ({ status: 200, body: {} }))
    const client = createClient({
      url: 'http://localhost:4440',
      organizationSlug: 'acme',
      fetch: fetchImpl,
    })
    await expect(
      client.oauth.handleOidcCallback(
        'https://app/auth/callback?error=access_denied',
      ),
    ).rejects.toThrow('OAuth error: access_denied')
  })
})

File: asgate/test/oauth_test.dart
import 'dart:convert';

import 'package:asgate/asgate.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:test/test.dart';

void main() {
  group('AsgateOAuthApi', () {
    test('getOidcSignInUrl parses provider and authorize_url', () async {
      final client = AsgateClient(
        url: 'https://auth.example.com',
        organizationSlug: 'my-org',
        httpClient: MockClient((req) async {
          expect(req.url.path, '/api/v1/auth/oidc/start');
          expect(req.headers['X-Organization-Slug'], 'my-org');
          return http.Response(
            jsonEncode({
              'status_code': 200,
              'message': 'ok',
              'data': {
                'provider': 'oidc_custom',
                'authorize_url':
                    'https://idp.example.com/authorize?client_id=x&state=s1',
              },
            }),
            200,
          );
        }),
      );
      final res = await client.oauth.getOidcSignInUrl();
      expect(res.provider, 'oidc_custom');
      expect(
        res.authorizeUrl,
        'https://idp.example.com/authorize?client_id=x&state=s1',
      );
      await client.dispose();
    });

    test('handleOidcCallback exchanges code and stores the session', () async {
      final client = AsgateClient(
        url: 'https://auth.example.com',
        organizationSlug: 'my-org',
        autoRefreshToken: false,
        httpClient: MockClient((req) async {
          expect(req.url.path, '/api/v1/auth/oidc/callback');
          expect(req.url.queryParameters['code'], 'c1');
          expect(req.url.queryParameters['state'], 's1');
          return http.Response(
            jsonEncode({
              'status_code': 200,
              'message': 'ok',
              'data': {
                'redirect_url':
                    'https://site.example.com/auth/callback'
                    '#access_token=at-1&refresh_token=rt-1',
              },
            }),
            200,
          );
        }),
      );
      final res = await client.oauth.handleOidcCallback(
        Uri.parse('myapp://auth/callback?code=c1&state=s1'),
      );
      expect(res.session, isNotNull);
      expect(res.session!.accessToken, 'at-1');
      expect(res.session!.refreshToken, 'rt-1');
      expect(client.isSignedIn, isTrue);
      expect(client.accessToken, 'at-1');
      await client.dispose();
    });

    test('handleOidcCallback throws on oauth error', () async {
      final client = AsgateClient(
        url: 'https://auth.example.com',
        organizationSlug: 'my-org',
      );
      await expectLater(
        client.oauth.handleOidcCallback(
          Uri.parse('myapp://cb?error=access_denied&state=s1'),
        ),
        throwsA(isA<AsgateException>()),
      );
      await client.dispose();
    });

    test('handleOidcCallback throws when code/state missing', () async {
      final client = AsgateClient(
        url: 'https://auth.example.com',
        organizationSlug: 'my-org',
      );
      await expectLater(
        client.oauth.handleOidcCallback(Uri.parse('myapp://cb')),
        throwsA(isA<AsgateException>()),
      );
      await client.dispose();
    });
  });
}
